### PR TITLE
[draft] Don't reset the RTC time from gettimeofday(), plus other RTC cleanups

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 

This is a fix for the RTC issues that [@sunnydaywest](https://forums.mbed.com/t/time-null-of-rtc-example-not-synchronizing-with-rtc-when-using-battery/22248) reported on the forums, and also as ARMmbed/mbed-os#15487.

As I explained in the thread, the easiest way to fix the issue is to update gettimeofday() to not reset the RTC time.  It's not intuitive for _getting_ the time to _reset_ the time, just because you didn't call rtc_init() first. 

While I was in the RTC code, I also took the opportunity to:
- Remove code from mbed_overrides.c which is also in rtc_api.c.  As far as we could tell, the code in rtc_api.c is the correct version, and it's what gets executed second so it makes sense for it to be the correct one.
- Move `target.lse_bypass` option to be for all STM32 targets, as this is a useful option which should work on all STM32 targets and might be wanted on custom boards
- Update docs for `target.clock_source` and `target.lse_drive_load_level` options

#### Impact of changes 

Calling time() or gettimeofday() without calling rtc_init() first will no longer cause the RTC time to be reset to 0.

### Documentation 

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [X] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers 

----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: targets/targets.json5 [hunk 1]
```diff
@@ -9763,7 +9764,7 @@
             "RTC"
         ]
     },
-    "RASPBERRY_PI_PICO": {
+    "RASPBERRY_PI_PICO_SWD": {
```
Comment (JohnK1987) [inline targets/targets.json5:9767]: This is not related to STM32 RTC

File: platform/source/mbed_rtc_time.cpp [hunk 2]
```diff
@@ -97,7 +97,7 @@ int gettimeofday(struct timeval *tv, MBED_UNUSED void *tz)
     _mutex->lock();
     if (_rtc_isenabled != NULL) {
         if (!(_rtc_isenabled())) {
-            set_time(0);
+            _rtc_init();
```
Comment (JohnK1987) [inline platform/source/mbed_rtc_time.cpp:100]: Just for sure. Are we sure this is a general API issue and not just STM32 related?

Comment (multiplemonomials) [reply platform/source/mbed_rtc_time.cpp:100]: Pretty sure, yeah!  It seems to me like this logic has always been wrong, we do not want to set the time to 0 in gettimeofday()!

**Comment by lefebvresam** [comment]
What would be the action to make time(NULL) working without calling rtc_init() first? It seems logical for me that RTC module is initialized when it's used and available and that you just can use the time(NULL) example as presented in the documentaion. Related to that, why you can use RTC at 4 different levels? As RTC object, RTC api, RTC HAL and default C-lib functions?

**Comment by JohnK1987** [comment]
From my understanding
* RTC HAL is brand specific HAL code in C. 
* Mbed hal is definition of how brand specific RTC HAL has to be represented in MbedOS in C
* Mbed RTC driver is wrapper from C to C++
You are able to use all layers in MbedOS.

After this PR the `rtc_init()` will be called inside of `gettimeofday()` and this is called inside of `time(NULL)` so finaly you can use it without extra call of `rtc_init()` or `set_time()`.

BR, Jan

**Comment by lefebvresam** [comment]
Is this for all functionality the same? It's strange that they use a mix in the mbed documentation. For instance for SPI they use the C++ way of instantiating objects.

File: targets/targets.json5 (hunk 1 above)
Comment (multiplemonomials) [reply targets/targets.json5]: Fixed!  This was a git error on my end X_X

**Comment by lefebvresam** [comment]
I will try this solution but I'm struggeling with (never used gh before):

```
sam@elitebook-845:~/git/hmc20_test/mbed-os$ gh pr checkout 210
could not determine base repo: none of the git remotes configured for this repository point to a known GitHub host. To tell gh about a new GitHub host, please use `gh auth login`
```

**Review by JohnK1987 — APPROVED** [review]

**Comment by JohnK1987** [comment]
> Is this for all functionality the same? It's strange that they use a mix in the mbed documentation. For instance for SPI they use the C++ way of instantiating objects.

I think yes and it is cool because Mbed can not cover every posibility and then you can combine Mbed APIs with brand specific HALs or make your own API based on Mbed HALs.

**Comment by lefebvresam** [comment]
Why this merge is blocked?

**Comment by multiplemonomials** [comment]
Huh, that is weird.  I am able to merge...

**Comment by multiplemonomials** [comment]
Oops looks like I forgot to remove "[draft]" from the title... oh well
